content/en/platform/corda/5.0/developing-applications/getting-started/utxo-advanced-ledger-extensions/commands-example.md

```diff
@@ -0,0 +1,243 @@
+
+```kotlin
+public final void verify(UtxoLedgerTransaction transaction) {
+  
+    List<? extends ExampleContractCommand> commands = transaction
+            .getCommands(ExampleContractCommand.class);
+            
+    for (ExampleContractCommand command : commands) {
+        if (command instanceof Create) verifyCreate(transaction);
+        else if (command instanceof Update) verifyUpdate(transaction);
+        else if (command instanceof Delete) verifyDelete(transaction);
+        else throw new IllegalStateException("Unrecognised command type.");
+    }
+}
+```
+
+The `verify` function is marked final for security reasons, and therefore additional commands cannot be added to the contract. For example, the contract may wish to describe multiple ways to `Update` a state, or set of states. The contract only defines a single `Update` command: there can only be one mechanism to perform updates.
+
+The second problem lies in the commands themselves and their names. `Create`, `Update` and `Delete` are very ambiguous names, which may not make sense depending on the context of the contract being implemented.
+
+## Delegated Command Design
+
+In the contracts above, the commands are nothing more than marker classes; effectively they are cases in a switch statement, which allows the contract's `verify` function to delegate responsibility of specific contract constraints to other functions, such as `verifyCreate`, `verifyUpdate` and `verifyDelete`.
+
+We can implement the `verify` function on the command itself. Instead of being an empty marker class, this gives the command responsibility, as it becomes responsible for implementing its associated contract verification constraints.
+In this case, we define a `VerifiableCommand` interface with a `verify` function; for example:
+
+```kotlin
+public interface VerifiableCommand extends Command {
+    void verify(UtxoLedgerTransaction transaction);
+}
+```
+
+Now that we have a command which itself can implement contract verification constraints, we can use this as the basis for the `ExampleContractCommand` class. This needs to be a class rather than an interface, because we need to be in complete control of its implementations for security.
+We achieve this by making the default constructor package private, so that only commands within the same package can extend it; for example:
+
+```kotlin
+ public class ExampleContractCommand implements VerifiableCommand {
+  ExampleContractCommand() { }
+}
+```
+
+Next, we can implement this interface as `Create`, `Update` and `Delete` commands; for example:
+
+```kotlin
+ public class Create extends ExampleContractCommand {
+    
+    @Override
+    
+    public final void verify(UtxoLedgerTransaction transaction) {
+    
+        // Verify base Create constraints
+        // Then verify additional Create constraints implemented in derived commands
+        onVerify(transaction);
+        
+    }
+    protected void onVerify(UtxoLedgerTransaction transaction) { }
+}
```

Comment:
```suggestion
```kotlin
public class Create extends ExampleContractCommand {
    @Override
    public final void verify(UtxoLedgerTransaction transaction) {
        // Verify base Create constraints
        // Then verify additional Create constraints implemented in derived commands
        onVerify(transaction);
    }
    
    protected void onVerify(UtxoLedgerTransaction transaction) { }
}
```